### PR TITLE
Add memberMap to State to speed up member queries

### PR DESCRIPTION
**Pull request by blackhole12**
This adds `memberMap` to the State object, in the same vein as the existing `guildMap` and `channelMap`, except it's a map of maps, mapping each guild ID to a map of member IDs paired to members. The map is replaced if necessary in `GuildAdd` so that all the pointers are always up-to-date, and is created in the `OnReady` function.

The Add, Remove, and Get functions have been updated to use the map, but still modify the slice if necessary, so the Member slice is still kept up to date. Speed tests indicate this approach is roughly 4-5 times faster for servers with 500+ members in them.

### CODE REVIEW

File: state.go [hunk 1]
```diff
@@ -77,6 +87,12 @@ func (s *State) GuildAdd(guild *Guild) error {
 		s.channelMap[c.ID] = c
 	}
 
+	// If this guild contains a new member slice, we must regenerate the member map so the pointers stay valid
+	if guild.Members != nil {
+		delete(s.memberMap, guild.ID)
```
Comment (iopred) [inline state.go]: Probably no benefit to deleting it first, it'll get overwritten in createMemberMap.

File: state.go [hunk 2]
```diff
@@ -312,18 +344,17 @@ func (s *State) Member(guildID, userID string) (*Member, error) {
...
 	s.RLock()
 	defer s.RUnlock()
 
-	for _, m := range guild.Members {
-		if m.User.ID == userID {
-			return m, nil
-		}
+	members, ok := s.memberMap[guildID]
+	if !ok {
+		return nil, errors.New("guild not found")
```
Comment (iopred) [inline state.go]: Return ErrNotFound here I guess.

**Comment by iopred** [comment]
Awesome, thanks for this PR! 💓

File: state.go (hunk 2 above)
Comment (blackhole12) [inline state.go]: This was debug code used to differentiate the guild not found error during testing, reverting it to ErrNotFound.

**Comment by blackhole12** [comment]
@iopred The test was adding a guild via GuildAdd() without having initialized it via OnReady, which wasn't properly handled. I added a check that initializes the member map if it doesn't exist in GuildAdd to ensure there is no possible case where it isn't initialized.

**Comment by iopred** [comment]
Thanks so much!

**Pull request merged**